kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -306,48 +318,92 @@ protected final Optional<List<FileStatus>> listDeltaAndCheckpointFiles(
+                                .orElse(true);
+                        if (!versionWithinRange) {
+                            // If we haven't taken any files yet and the first file we see is
+                            // greater than the versionToLoad then the versionToLoad is not
+                            // reconstructable from the existing logs
+                            if (output.isEmpty()) {
+                                long earliestVersion = DeltaHistoryManager
+                                        .getEarliestRecreatableCommit(engine, logPath);
+                                throw DeltaErrors.versionBeforeFirstAvailableCommit(
+                                        tablePath.toString(), versionToLoad.get(), earliestVersion);
+                            }
+                            break;
+                        }
 
-                    // Take files until the version we want to load
-                    final boolean versionWithinRange = versionToLoad
-                        .map(v -> FileNames.getFileVersion(new Path(fileStatus.getPath())) <= v)
-                        .orElse(true);
-
-                    if (!versionWithinRange) {
-                        // If we haven't taken any files yet and the first file we see is greater
-                        // than the versionToLoad then the versionToLoad is not reconstructable
-                        // from the existing logs
-                        if (output.isEmpty()) {
-                            long earliestVersion = DeltaHistoryManager.getEarliestRecreatableCommit(
-                                engine, logPath);
-                            throw DeltaErrors.versionBeforeFirstAvailableCommit(
-                                tablePath.toString(), versionToLoad.get(), earliestVersion);
+                        // Ideally listFromOrNone should return lexiographically sorted files amd so
+                        // maxDeltaVersionSeen should be equal to fileVersion.
+                        // But we are being defensive here and taking max of all the fileVersions
+                        // seen.
+                        if (FileNames.isCommitFile(fileName)) {
+                            maxDeltaVersionSeen.set(Math.max(maxDeltaVersionSeen.get(),
+                                    FileNames.deltaVersion(fileStatus.getPath())));
                         }
-                        break;
+
+                        output.add(fileStatus);
                     }
 
-                    output.add(fileStatus);
-                }
+                    return output;
+                });
 
-                return output;
-            });
+        if (!commitCoordinatorClientHandlerOpt.isPresent()) {
+            return resultFromFsListingOpt;
+        }
+
+        List<FileStatus> unbackfilledCommitsFiltered = new ArrayList<>();
+        boolean dropConditionMet = false;
+        for (Commit commit : unbackfilledCommits) {
```

Comment:
Would this work?
```
List<FileStatus> relevantUnbackfilledCommits = unbackfilledCommits
 .stream()
 .filter((commit) -> commit.getVersion() > maxDeltaVersionSeen.get())
 .takeWhile((commit) -> !versionToLoad.isPresent() || commit.getVersion() <= versionToLoad.get())
 .toArray()
```

Reply:
The `takeWhile` API is since 9+. So I use `filter` instead.